[Nicholas-Cheng-De-Fei] ip

df-Nic opened the pull request:
# Quack

Come and use this ~chatbot~ DUCK to help **keep track of meaningful tasks**!

Why use a duck? Well :
- I am fast
- I have unlimited memory (_I hope_)
- I am very easy to use

## How to use me

Step 1: **Download the zip file**
> On the [GitHub repository](https://github.com/Nicholas-Cheng-De-Fei/ip) download the code as a Zip File

Step 2: **Unzip the file**
> Extract the files in the zip file to wherever you want to keep this duck on your laptop

Step 3: **Run!!**
> Using any IDE, run the `Quack.java` file to start me!

## Features
- [x] **Add Tasks**
- [x] **Delete Tasks**
- [x] **Mark/Unmark Tasks**
- [x] **Find Tasks**
- [x] **Save Tasks to Memory**
- [x] **GUI**
- [x] **Help feature (_In progress_)**

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

Commit:
Renamed chatbot to Quack and added functions for the bot to greet and exit

Commit:
Chatbot can now echo user input and will exit when the user inputs bye

Commit:
Replaced echo feature with an item storing feature, allowing users to store texted that they key in

Commit:
ToDoList now stores an array of Tasks, which can be checked instead of just string

Commit:
Modify the chatbot such that it will only take in certain commands (list and bye) for now. If the user inputs an invalid command it will promt the user that the command does not exist

Commit:
Added functionality to mark and unmark tasks. Also added a stand-alone command to add tasks to the list

Commit:
Changed the class name from ToDoList to TaskList

Commit:
Added a ToDoTask class for ToDo tasks

Commit:
Added the deadline task class to add deadlines to the list

Commit:
Added the functionality for the user to add a deadlines

Commit:
The chatbot can handle invalid inputs

Commit:
Chatbot can handle invalid inputs when adding tasks, for instance invalid task type or empty description

Commit:
Added the functionality to delete a task from the lst

Commit:
Added input validation for delete function and additional functions like mark and unmark

Commit:
Fixed a typo in the code

Commit:
Updated the README file

Commit:
Added Text UI Testing

Commit:
Added enum TaskType to define the types of task the chatbot can store

Commit:
Fix line length in TaskList.java and InvalidTaskTypeException.java

Commit:
Update content in README.md file

Commit:
Add Javadoc to Quack.java

Commit:
Add JavaDoc to TaskList.java

Commit:
Add JavaDoc to Task.java

Commit:
Add JavaDoc for the remainding java files

Commit:
Update JavaDoc to all .java files

Commit:
Add save task functionality for Quack

Commit:
Add functionality to load tasks from the save file

Commit:
Merge branch 'branch-Level-7'

Commit:
Update JavaDoc in Test.java and TaskList.java

Commit:
Change DeadlineTask.java and EventTask.java to use LocalDate instead of String

Commit:
Modify Quack to take in date time inputs for events and deadlines

Commit:
Impliment input validation for date time inputs

Commit:
Merge branch 'branch-Level-8'

Commit:
Update JavaDoc in Quack.java

Commit:
Add Storage class to handle saving and loading

Commit:
Add Parser class to handle user input

Commit:
Add Ui class to handle user interaction

Commit:
Merge branch 'branch-A-MoreOOP'

Commit:
Package classes into their respective packages

Commit:
Import packages to all the relevant .java files

Commit:
Merge branch 'branch-A-Packages'

Commit:
Move code to another folder

Commit:
Merge remote-tracking branch 'origin/add-gradle-support' into branch-A-Gradle

Commit:
Add JUnit Testing

Commit:
Merge branch 'branch-A-Gradle'

Commit:
Add more test cases

Commit:
Merge branch 'branch-A-JUnit'

Commit:
Update JavaDocs

Commit:
Refactor the code to follow coding standards

Commit:
Update JavaDoc to follow the standard convention

Commit:
Add find feature to Quack

Commit:
Add JUnit test cases for the find function

Commit:
Merge branch 'branch-Level-9'

Commit:
Update conding standards

Commit:
Merge branch 'branch-A-CodingStandard'

Commit:
Merge branch 'branch-A-JavaDoc'

Commit:
Change code to following coding standards

colinac commented on src/main/java/quack/Paser.java:
```diff
@@ -0,0 +1,92 @@
+/**
+ * This class is responsible for handling user input commands.
+ */
+public class Paser {
```
Perhaps you meant parser instead?

df-Nic replied:
Thanks for pointing this typo out, I have already changed the class name in my recent commits.

colinac commented on src/main/java/quack/Ui.java:
```diff
@@ -0,0 +1,252 @@
+        "\\_____\\ \\_/____/(____  /\\___  >__|_ \\ \n" +
+        "       \\__>          \\/     \\/     \\/\n";
+
+    /** Farewell message for Quack */
```
I like how you gave comments for readers to understand the intent of each variable.

colinac commented on src/main/java/quack/Ui.java:
```diff
@@ -0,0 +1,252 @@
+public class Ui {
+    
+    /** String to print out the spacers between each command */
+    private String spacer = "-".repeat(65);
```
Add a constant to avoid magic numbers (65 in this case)?

df-Nic replied:
Thanks for spotting this out. I have removed this line of code after implementing JavaFX

colinac commented on src/main/java/quack/command/Command.java:
```diff
@@ -0,0 +1,12 @@
+ * This is an abstract class to define a command object.
+ */
+public abstract class Command {
+
```
I like how you added the command class which makes the code extensible 👍

colinac reviewed (commented):
The code has implemented OOP nicely and tests each command rather thoroughly. Great job!

Commit:
Add checkstyle to Quack project

Commit:
Merge branch 'branch-A-CheckStyle'

Commit:
Change code to follow coding standards

Commit:
Add basic GUI to Quack project

Commit:
Modify commands to work with JavaFX

Commit:
Modify AddTaskCommand.java to work with JavaFX

Commit:
Merge branch 'branch-Level-10'

Commit:
Update Task.java factory method to use varargs

Commit:
Merge branch 'branch-A-Varargs'

Commit:
Update JavaDoc

DINESH1201 commented on README.md:
I like how your readme encompasses all the features of your chat bot nicely.

DINESH1201 commented on src/main/java/quack/Storage.java:
```diff
@@ -0,0 +1,141 @@
+        try {
+            Path filePath = Paths.get(DEFAULT_PATH);
+            Files.createDirectories(filePath);
+        } catch (Exception error) {
```
Perhaps catching the FileNotFound error or whatever error is thrown is better here

DINESH1201 commented on src/main/java/quack/command/Command.java:
```diff
@@ -0,0 +1,40 @@
+/**
+ * This is an abstract class to define a command object.
+ */
+public abstract class Command {
```
I like that you made the command class into an abstract class. This would really help in the future when you need to add more commands/features.

DINESH1201 reviewed (commented):
I like how comprehensive your code is

Commit:
Add JUnit tests for Todo and Deadline tasks

Commit:
Add JUnit tests for the EventTask class

df-Nic committed:
Merge pull request #1 from Nicholas-Cheng-De-Fei/branch-JUnit

Branch JUnit

Commit:
Rewrite code in test file to follow coding standards

Coding standards: make test files following the coding standards.

To ensure code quality, the code must follow the coding standards.

Using the checkstyle tools by gradle, any lines of code that violates
the coding standards will be flagged and they will then be changed
accordingly.

Using an automated style cheker saves time and it can also spot errors
which we might not detect through manual work.

Commit:
Add Java assertions to project

Command classes: add Java assertions

Java assertions helps to identify potential bugs within the code base by
stating assumptions about the logic of the code.

Java assertions are added at certain points within the function mainly
at the start of the function to ensure that the input is not null.

These assertions are placed at this specfic points is because if the
input is null it means that the program is not capturing and passing the
input correctly to the respective function.

Commit:
Update build.gradle to enable assertions

Commit:
Change toString funcrion in TaskList.java to use streams

Commit:
Improve code quality of the command classes

Commit:
Configure code to follow coding standards

df-Nic committed:
Merge pull request #2 from Nicholas-Cheng-De-Fei/branch-A-CodeQuality

Branch a code quality

Commit:
Merge branch 'master' into branch-A-Streams

df-Nic committed:
Merge pull request #3 from Nicholas-Cheng-De-Fei/branch-A-Streams

Change toString funcrion in TaskList.java to use streams

Commit:
Edit logic in AddTaskCommand.java

Commit:
Merge branch 'branch-A-Assertions'

Commit:
Add assert line in Task.java

Commit:
Add assert statement in TaskList.java

df-Nic committed:
Merge pull request #5 from Nicholas-Cheng-De-Fei/branch-A-Assertions

Add assert statement in TaskList.java

Commit:
Add gradle.yml file for CI

df-Nic committed:
Merge pull request #6 from Nicholas-Cheng-De-Fei/branch-A-CI

Add gradle.yml file for CI

Commit:
Add tag command to Quack

Tag Command: allows assigning of tags to tasks

Allow rhe user to better organize their tasks.

Add a new command tag which will prompt the user for an index and a tag
label and then add the label to the task.
There us also an untag command which will prompt the user for an
index and then remove the tag from the task.

Asking the user for the index of where the task is located will make
finding easier. And as for prompting the user for a tag is to allow
freedom to choose the taging system the user wants to impliment.

Commit:
Modify Storage.java to handle tags

df-Nic committed:
Merge pull request #7 from Nicholas-Cheng-De-Fei/branch-BCD-Extension

Branch bcd extension

Commit:
Move gradle.yml file to the correct folder

Commit:
Update gardle.yml file

Commit:
Add more personality in Quack chatbot responses

Commit:
Add date time checks when creating a task

Commit:
Update README.md

Commit:
Update UI for Quack

Commit:
Change input field background colour

df-Nic committed:
Merge pull request #9 from Nicholas-Cheng-De-Fei/branch-A-BetterGui

Branch a better gui

Commit:
Update code to follow coding standards

df-Nic committed:
Merge pull request #11 from Nicholas-Cheng-De-Fei/branch-A-Personality

Branch a personality

df-Nic committed:
Merge pull request #12 from Nicholas-Cheng-De-Fei/branch-A-MoreErrorHandling

Add date time checks when creating a task

Commit:
Add reference to usage of AI assisted tools

Commit:
Edit AI.md file

df-Nic committed:
Merge pull request #13 from Nicholas-Cheng-De-Fei/branch-A-AiAssisted

Branch a ai assisted

Commit:
Add equals function in task classes

Commit:
Add JUnit testing for Parser and TaskList classes

Commit:
Update Parser unit test

Commit:
Update ParserTest.java to follow coding standards

Commit:
Remove stub files

df-Nic committed:
Merge pull request #15 from Nicholas-Cheng-De-Fei/branch-A-MoreTesting

Branch a more testing

Commit:
Add handling of blank or whitespace only task description

df-Nic committed:
Merge pull request #16 from Nicholas-Cheng-De-Fei/branch-ErrorHandling

Add handling of blank or whitespace only task description

Commit:
Update README.md file

Commit:
Add and update test cases

df-Nic committed:
Merge pull request #17 from Nicholas-Cheng-De-Fei/branch-JUnitTesting

Add and update test cases

Commit:
Add help command for Quack

Commit:
Add test casses for the help command

df-Nic committed:
Merge pull request #18 from Nicholas-Cheng-De-Fei/branch-HelpCommand

Branch help command

Commit:
Update code to improve on code quality

Commit:
Change javafx version from 22 to 17

Commit:
Update README.md in docs folder

Commit:
Add user guide to Quack

Commit:
Comment out user guide

Commit:
Modify user guide

Commit:
Fix formatting issues in the user guide

Commit:
Fix list formatting issue

Commit:
Update formating on user guide

Commit:
Update formatting for images in the user guide

df-Nic committed:
Merge pull request #19 from Nicholas-Cheng-De-Fei/branch-A-UserGuide

Branch a user guide

Commit:
Update release link in the user guide

Commit:
Update user guide markdown file

Commit:
Update list in README.md for the user guide

Commit:
Update the user guide

Commit:
Update parser command to trim excess whitespaces

Commit:
Add more information to the user guide for quack
